### ISSUE TITLE
rewrite_query API stability/programmatic setting

### ISSUE THREAD

**Issue by jheld** [adamchainz/django-mysql#643]
Summary: *Consider making rewrite setting/flow programmatic*

`DJANGO_MYSQL_REWRITE_QUERIES` setting is static. We do understand why that's the case historically (powerful but easy to turn off). After reading through the code it seems that the initial small overhead to each query is a string comparison to check for the rewrite marker in the sql, when the setting is on. Although the string check generally should be quite fast (sql queries in mysql in general cannot go above 1GB anyway, and most likely are much smaller than that), the overhead does indeed still pose a performance risk for larger queries (and it's not always known if/when they could occur).

We are considering attempting to add support to make the setting allowed to be a callable. So for instance, if we were to make a django request middleware (considering certain user permissions, or certain APIs), we could make the setting on/off during the execution of that particular request.

Now, having read the source code for rewrite_query and the monkey_patches, I'm not certain that a middleware would be effective (perhaps you can verify my question).

On the other hand, middleware or not, if we could guarantee a sort of middle level public API for what the monkey patching does (checks the setting but also alters the sql if it matches the rewrite logic, based on wrapping the cursor's execute function), then that potentially could allow a ton of official flexibility for projects that wish to have programmatic access to rewrite, if they so choose, and without risking a breakage in compatibility with this library should something change (since it is only a somewhat exposed function, incidentally public, rather than explicitly).

There may be methods to hack around this (perceived limitation), but I am hoping that we can find a more official path that embraces flexibility, rather than being unaware that projects might be relying on implied behavior.

What are your thoughts so far?

**Comment by adamchainz** [comment]
> the overhead does indeed still pose a performance risk for larger queries (and it's not always known if/when they could occur).

First, I'll address this. I don't believe there's much overhead, and you haven't provided any evidence that it's noticeable. If you do any profiling I think you'll find the overhead imperceptible compared to many other parts of the Django ORM.

Searching for the marker in a 1000 character string is 200 nanoseconds:

```
In [1]: string = 1_000 * "B"

In [2]: marker ="/*QueryRewrite':"

In [3]: %timeit marker in string
207 ns ± 0.796 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Many queries are even shorter so the overhead will be less. Yes, there's a settings lookup and a function call missing from this profile, but that should be minimal.

Additionally, there are many more low-hanging fruit for optimization in Djagno core. For example take this PR I made *this week*: https://github.com/django/django/pull/12772 . This saved a regex compile on *every single* query. On my machine the saving there is 2 microseconds:

```
In [5]: %timeit re.compile(r'^(.*)\s(ASC|DESC)(.*)', re.MULTILINE | re.DOTALL)
2.09 µs ± 127 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Basically 10x the overhead of rewrite queries. And this is a relatively small optimization - see e.g. https://github.com/django/django/pull/12337 which saved ~1% of test suite runtime through some relatively simple caching of model properties.

Please use a profiler (I like https://github.com/benfred/py-spy ) to find where to work on performance, we can use your help!

> On the other hand, middleware or not, if we could guarantee a sort of middle level public API for what the monkey patching does

Moving to database instrumentation would allow temporary installation of the rewriting logic. I thought I had an issue for that but I don't; it's here now: #644.
